apis/compute/v1beta1/network_reference.go

```diff
@@ -0,0 +1,192 @@
+func (r *ComputeNetworkRef) ValidateExternal(ref string) error {
+	id := &NetworkIdentity{}
+	if err := id.FromExternal(r.GetExternal()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (r *ComputeNetworkRef) Normalize(ctx context.Context, reader client.Reader, defaultNamespace string) error {
+	if r.External != "" {
+		_, err := ParseComputeNetworkExternal(r.External)
+		if err != nil {
+			return err
+		}
+		external := common.FixStaleComputeExternalFormat(r.External)
+		r.External = external
+		return nil
+	}
+
+	// From the Config Connector object
+	if r.Namespace == "" {
+		r.Namespace = defaultNamespace
+	}
+	key := types.NamespacedName{Name: r.Name, Namespace: r.Namespace}
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(ComputeNetworkGVK)
+	if err := reader.Get(ctx, key, u); err != nil {
+		if apierrors.IsNotFound(err) {
+			return k8s.NewReferenceNotFoundError(u.GroupVersionKind(), key)
+		}
+		return fmt.Errorf("reading referenced %s %s: %w", ComputeNetworkGVK, key, err)
+	}
+
+	// Get external from status.externalRef. This is the most trustworthy place.
+	actualExternalRef, _, err := unstructured.NestedString(u.Object, "status", "externalRef")
+	if err != nil {
+		return fmt.Errorf("reading status.externalRef: %w", err)
+	}
+	if actualExternalRef != "" {
+		r.External = actualExternalRef
+		return nil
+	}
+
+	// Get external from status.selfLink. This ensures backward compatibility for TF/DCL-based resources that lack status.externalRef.
+	selfLink, _, err := unstructured.NestedString(u.Object, "status", "selfLink")
+	if err != nil {
+		return fmt.Errorf("reading status.selfLink: %w", err)
+	}
+	if selfLink == "" {
+		return k8s.NewReferenceNotFoundError(u.GroupVersionKind(), key)
+	}
+
+	external := common.FixStaleComputeExternalFormat(selfLink)
+	r.External = external
+	return nil
+}
+
+func (id *NetworkIdentity) ConvertToProjectNumber(ctx context.Context, projectMapper *projects.ProjectMapper) error {
+	if id == nil {
+		return nil
+	}
+
+	projectNumber, err := projectMapper.LookupProjectNumber(ctx, id.Parent().ProjectID)
+	if err != nil {
+		return fmt.Errorf("error looking up project number for project %q: %w", id.Parent().ProjectID, err)
+	}
+
+	id.parent.ProjectID = strconv.FormatInt(projectNumber, 10)
+	return nil
+}
+
+// ConvertToProjectNumber converts the external reference to use a project number.
+func (ref *ComputeNetworkRef) ConvertToProjectNumber(ctx context.Context, projectMapper *projects.ProjectMapper) error {
+	if ref == nil {
+		return nil
+	}
+
+	id, err := ParseComputeNetworkExternal(ref.External)
+	if err != nil {
+		return err
+	}
+
+	if err := id.ConvertToProjectNumber(ctx, projectMapper); err != nil {
+		return err
+	}
+
+	ref.External = id.String()
+	return nil
+}
+
+// ConvertClientToProjectNumber converts the external reference to use a project number.
+func (ref *ComputeNetworkRef) ConvertClientToProjectNumber(ctx context.Context, projectsClient *resourcemanager.ProjectsClient) error {
```

Comment:
nit: i'm fine to have this live here but it feels more like a apis/common thing?